docs/src/proposals/multi-writers.md

```diff
@@ -0,0 +1,188 @@
+received field is updated by taking the max of the current index + 1 and the
+existing value [here](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3180-L3182).
+To correctly handle a potential data race where the later smaller value
+overrides the older bigger value, a merge operator that implements such max
+operation should be used.
+
+#### [first_shred_timestamp](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L24-L25)
+Similar to the received field, `first_shred_timestamp` can use merge operator
+that includes min operation to correctly capture the `first_shred_timestamp` to
+avoid the case where more than one writers are eligible for updating the
+`first_shred_timestamp` field to avoid potential override in case we want to
+keep this field accurate.
+
+#### [last_index](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L26-L27)
+The last_index field doesn't require merge operator to ensure its atomicity as
+there should be only one shred which has its [is_last_in_slot to be true]
+(https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3189-L3200).
+
+#### [parent_slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L28-L29)
+parent_slot is updated only from [std::u64::MAX to a valid parent slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3251-L3253).
+A merge operator is needed to ensure we never override a valid parent slot
+by u64::MAX.
+
+#### [next_slots](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L30-L32)
+next_slot is updated under [`handle_chaining`](https://github.com/yhchiang-sol/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L933)
+inside [`chain_new_slot_to_prev_slot`](https://github.com/yhchiang-sol/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3543-L3550).
+As we never remove entries from the next_slots field, its merge operator can be
+implemented by taking the union of the new value and the existing value.
+
+#### [is_connected](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L33-L35)
+`is_connected` field is either updated based on
+[the previous slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3535)
+or via [propagation](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3469).
+As `is_connected` is never set from true to false, its merge operator can be
+implemented using the OR operation.
```

Comment:
heh this field is already so busted and unused I'm temped to remove it and all the complicated associated logic. I think it's probably broken in the majority of validators that started from a snapshot. 

The reason being when you start from a snapshot of slot `X`, you unpack all the account state for slot `X`, but `X` does not actually exist in the ledger (You don't need the transactions from `X` since the snapshot already contains all of the state). 

Then all blockstore slots will descend from `X` because it's the root block, so blockstore insert will insert an empty `SlotMeta` for `X` whenever it inserts a shred from a child slot of `X` (kind of like what we do now if you get a shrd from the child slot before the parent slot). This empty `SlotMeta` for `X` will have `is_connected == false`, and so all children will have `is_connected == false` as well